core/src/test/java/org/mule/el/mvel/datatype/AbstractVarAssignmentDataTypePropagatorTestCase.java

```diff
@@ -53,16 +57,42 @@ protected void doAssignmentDataTypePropagationTest(PropertyScope scope, String e
 
         MuleEvent testEvent = getTestEvent(TEST_MESSAGE);
 
+        CompiledExpression compiledExpression = compileMelExpression(expression, testEvent);
+
+        dataTypePropagator.propagate(testEvent.getMessage(), new TypedValue(TEST_MESSAGE, expectedDataType), compiledExpression);
+
+        assertThat(testEvent.getMessage().getPropertyDataType(PROPERTY_NAME, scope), like(String.class, JSON, CUSTOM_ENCODING));
+    }
+
+    protected void doInnerAssignmentDataTypePropagationTest(PropertyScope scope, String expression) throws Exception
+    {
+        final DataType expectedDataType = DataTypeFactory.create(Map.class, UNKNOWN);
+        expectedDataType.setEncoding(CUSTOM_ENCODING);
+
+        MuleEvent testEvent = getTestEvent(TEST_MESSAGE);
+        final Map<String, String> propertyValue = new HashMap<>();
+        propertyValue.put(INNER_PROPERTY_NAME, TEST_MESSAGE);
+        testEvent.getMessage().setProperty(PROPERTY_NAME, propertyValue, scope, expectedDataType);
+
+        CompiledExpression compiledExpression = compileMelExpression(expression, testEvent);
+
+        // Attempts to propagate a different dataType, which should be ignored
+        dataTypePropagator.propagate(testEvent.getMessage(), new TypedValue(propertyValue, DataType.STRING_DATA_TYPE), compiledExpression);
```

Comment:
There are two datatypes involved. One is the map dataType which is type of the property in the message. Another one is the String dataType which is the dataType of the value which is enriched with the map expression.
This should not fail at all, is just checking that the propagation didn't affect the dataType in the property